components/compliance-service/profiles/db/store.go

```diff
@@ -376,89 +379,108 @@ func (s *Store) Delete(namespace string, name string, version string) error {
-		`, sort, order)
-
-		rows, err = s.DB.Query(query)
-	case len(namespace) == 0:
-		query := fmt.Sprintf(`
-			SELECT sha256, info
-			FROM store_profiles WHERE sha256 IN
-			(SELECT sha256 FROM store_market)
-			AND info->>'name' = $1
-			ORDER BY store_profiles.info #>> '{%s}' %s, store_profiles.info #>> '{version}' DESC;
-		`, sort, order)
-
-		rows, err = s.DB.Query(query, name)
-	case len(name) == 0:
-		query := fmt.Sprintf(`
-			SELECT sha256, info
-			FROM store_profiles WHERE
-			sha256 IN (SELECT sha256 FROM store_namespace WHERE owner=$1)
-			ORDER BY store_profiles.info #>> '{%s}' %s, store_profiles.info #>> '{version}' DESC;
-		`, sort, order)
-
-		rows, err = s.DB.Query(query, namespace)
-	default:
-		query := fmt.Sprintf(`
-			SELECT sha256, info
-			FROM store_profiles WHERE
-			sha256 IN (SELECT sha256 FROM store_namespace WHERE owner=$1)
-			AND info->>'name' = $2
-			ORDER BY store_profiles.info #>> '{%s}' %s, store_profiles.info #>> '{version}' DESC;
-		`, sort, order)
-
-		rows, err = s.DB.Query(query, namespace, name)
+type ProfilesListRequest struct {
+	Filters   map[string][]string
+	Name      string
+	Namespace string
+	Order     string
+	Page      int
+	PerPage   int
+	Sort      string
+}
+
+func (s *Store) ListProfilesMetadata(req ProfilesListRequest) ([]inspec.Metadata, int, error) {
+	if req.Order != "ASC" && req.Order != "DESC" {
+		return nil, 0, status.Errorf(codes.InvalidArgument, "order field '%s' is invalid. Use either 'ASC' or 'DESC'", req.Order)
+	}
+
+	if req.Sort != "name" && req.Sort != "title" && req.Sort != "maintainer" {
+		return nil, 0, status.Errorf(codes.InvalidArgument, "sort field '%s' is invalid. Use either 'name', 'title' or 'maintainer'", req.Sort)
+	}
+
+	sql := squirrel.
+		Select("sha256, info, count(*) OVER() AS total").
+		From("store_profiles")
+
+	if len(req.Namespace) == 0 {
+		sql = sql.
+			Where("exists(select 1 from store_market where store_profiles.sha256 = store_market.sha256)")
+	} else {
+		sql = sql.
+			Where("exists(select 1 from store_namespace where store_profiles.sha256 = store_namespace.sha256 and store_namespace.owner = ?)", req.Namespace)
+	}
+
+	if len(req.Name) != 0 {
+		sql = sql.Where("info->>'name' = ?", req.Name)
 	}
+
+	for key, values := range req.Filters {
+		terms := squirrel.Or{}
+		for _, value := range values {
+			term := fmt.Sprintf("info->>'%s' like ?", key)
+			v := pgutils.EscapeLiteralForPGPatternMatch(value)
+			v = strings.ReplaceAll(v, "*", "%")
+			terms = append(terms, squirrel.Expr(term, v))
+		}
+		sql = sql.Where(terms)
+	}
+
+	// For backwards compatibility, pagination needs to be optional. Including a value for PerPage is how a caller can opt-in to pagination.
+	if req.PerPage > 0 {
+		perPage := uint64(req.PerPage)
+		page := uint64(req.Page)
+
+		sql = sql.
+			Limit(perPage).
+			Offset(perPage * page)
+	}
+
+	sql = sql.OrderBy(fmt.Sprintf("info->>'%s' %s", req.Sort, req.Order), "info->>'version' DESC")
```

Comment:
The latest released version of squirrel only has `OrderBy`, which only accepts entire order by clauses. On master there's a new function `OrderByClause` which will accept parameters so we could potentially avoid the `Sprintf` call here: https://github.com/Masterminds/squirrel/blob/d13326f0be730e8aa260f440b63fd00c858a03d0/select.go#L330-L342